packages/datadog-instrumentations/src/aws-sdk.js

```diff
@@ -0,0 +1,69 @@
+'use strict'
+
+const {
+  channel,
+  addHook,
+  AsyncResource
+} = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+
+const startCh = channel('apm:aws:request:start')
+const responseCh = channel('apm:aws:response')
+const completeCh = channel('apm:aws:request:complete')
+
+function wrapRequest (send) {
+  return function wrappedRequest (cb) {
+    if (!this.service) return send.apply(this, arguments)
+    const outerAr = new AsyncResource('apm:aws:request:outer')
+
+    const serviceIdentifier = this.service.serviceIdentifier
+
+    this.on('complete', response => {
+      completeCh.publish({ response, serviceIdentifier })
+    })
+
+    return new AsyncResource('apm:aws:request:inner').runInAsyncScope(() => {
+      startCh.publish({
+        serviceIdentifier,
+        operation: this.operation,
+        awsRegion: this.service.config && this.service.config.region,
+        awsService: this.service.api && this.service.api.className,
+        request: this
+      })
+
+      if (typeof cb === 'function') {
+        arguments[0] = wrapCb(cb, serviceIdentifier, this, outerAr)
+      }
+      return send.apply(this, arguments)
+    })
+  }
+}
+
+function wrapCb (cb, serviceName, request, ar) {
+  return function wrappedCb (err, response) {
```

Comment:
The error should be emitted.

Reply:
Also, is there always a response even when there is an error?

Reply:
> The error should be emitted.

This is handled here, noting that the response object has an error property. https://github.com/DataDog/dd-trace-js/pull/2028/files/aabb5e3eb6f0b83852ec99b04b2655dd653a115a#diff-ca0461c83e8fdabf39ccf054f5e3754d3d9236901cc96bb02e8c0cdbb994335bR21-R23

> Also, is there always a response even when there is an error?

Yes. It's even attached to the response object.

Reply:
Could it still be emitted alone to avoid relying on implementation details? Not sure if that makes sense as I haven't looked into it, but something to consider.